optuna/integration/mlflow.py

```diff
@@ -234,12 +247,17 @@ def _set_tags(self, trial: optuna.trial.FrozenTrial, study: optuna.study.Study)
         directions = [d.name for d in study.directions]
         tags["direction"] = directions if len(directions) != 1 else directions[0]
 
-        tags.update(trial.user_attrs)
         distributions = {(k + "_distribution"): str(v) for (k, v) in trial.distributions.items()}
-        tags.update(distributions)
+        study_attrs = study.user_attrs
+        trial_attrs = trial.user_attrs
+        collisions = study_attrs.keys() & trial_attrs.keys()
+
+        for key in collisions:
+            # make colliding trial and study user attributes unique
+            study_attrs["study_{}".format(key)] = study_attrs.pop(key)
+            trial_attrs["trial_{}".format(key)] = trial_attrs.pop(key)
```

Comment:
The duplicated attribute keys are implicitly renamed here. I'm concerned about the following two points:

- Users may not find the renamed attributes. This behavior is not documented.
- The renamed keys can collide with the other keys

For the second point, `study_foo` in trial's user attributes are overwritten by the value of `foo` in study's user attributes.

```python
import optuna
from optuna.integration.mlflow import MLflowCallback


def objective(trial):
    x = trial.suggest_float("x", -10, 10)
    trial.set_user_attr("foo", "baz")
    trial.set_user_attr("study_foo", "baz")
    return (x - 2) ** 2


mlflc = MLflowCallback(
    tracking_uri=None,
    metric_name="my metric score",
    mlflow_kwargs={}
)

study = optuna.create_study()
study.set_user_attr("foo", "bar")
study.optimize(objective, n_trials=10, callbacks=[mlflc])
```

I may be conservative, but the current master's behavior looks intuitive to me.

Reply:
My [reason to remove](https://github.com/optuna/optuna/pull/2912#discussion_r704658783) `tag_study_user_attrs` was based on different behavior when trial/study user attributes are logged (one conditionally, other one always), which I found counter intuitive. Collisions can happen in both configurations, but probably are rare cases, as @nzw0301 noted.

Anyway I'm fine to revert those changes to master.

Reply:
> different behavior when trial/study user attributes are logged (one conditionally, other one always)

Thank you for your explanation.
Hmm, if the mlflow has a target of tags corresponding to Optuna's study, the situation would be simple.

I guess users who add many user attributes to studies may want to turn off the record of studys' user attributes to keep the mlflow experiment clean. The `tag_study_user_attrs` flag will help such users.

> Collisions can happen in both configurations, but probably are rare cases, as @nzw0301 noted.

I agree with you in terms of the frequency. On the other hand, Optuna is getting more and more users, and such rare cases tend to be reported to the issues and gitter. Currently, the priority is written in the document, and users can resolve issues by themselves.

> Please note that when this flag is set, key value pairs in study.user_attrs will supersede existing tags.

https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.MLflowCallback.html#optuna.integration.MLflowCallback 

> Anyway I'm fine to revert those changes to master.

@nzw0301 What do you think of it?

Reply:
Thank both of you for your comments. 

> The renamed keys can collide with the other keys
Indeed, I missed this point.

Users might also use `trial.user_attrs` for such a rate case. So if we do not save `study.user_attr` automatically, both `trial.user_attr` and `study.user_attr` should be conditional as suggested in https://github.com/optuna/optuna/pull/2912#discussion_r704658783.

Reply:
@nzw0301 
> So if we do not save study.user_attr automatically, both trial.user_attr and study.user_attr should be conditional as suggested in #2912 (comment).

Thank you for your comment. Sounds reasonable.

By the way, I think introducing `mlflow_kwargs` and handling of `trial.user_attrs` and `study.user_attrs` are separable development items. How about focusing on the `mlflow_kwargs` argument in this PR and creating a new PR for user attributes?

Reply:
>By the way, I think introducing mlflow_kwargs and handling of trial.user_attrs and study.user_attrs are separable development items. How about focusing on the mlflow_kwargs argument in this PR and creating a new PR for user attributes?

Makes sense, let me revert changes to `*user_attrs` for now.

Reply:
Done in b822511b391ffb8a1c580b1873fc94594c9f4c22.